libbpf-rs/src/object.rs

```diff
@@ -181,25 +188,36 @@ impl OpenObject {
     ///
     /// # Safety
     ///
-    /// If `ptr` is unopen or already loaded then further operations on the returned object are
-    /// undefined.
+    /// Operations on the returned object are undefined if `ptr` is any one of:
+    ///     - null
+    ///     - point to an unopened bpf_object
+    ///     - point to a loaded
     ///
     /// It is not safe to manipulate `ptr` after this operation.
     pub unsafe fn from_ptr(ptr: *mut libbpf_sys::bpf_object) -> Result<Self> {
-        Self::new(ptr)
+        unsafe { Self::new(NonNull::new_unchecked(ptr)) }
     }
 
     /// Takes underlying `libbpf_sys::bpf_object` pointer.
-    pub fn take_ptr(mut self) -> *mut libbpf_sys::bpf_object {
-        let ptr = self.ptr;
-        self.ptr = ptr::null_mut();
+    pub fn take_ptr(mut self) -> NonNull<libbpf_sys::bpf_object> {
+        let ptr = {
+            // manually free the internal state.
+            // using destructuring we make sure we'll get a compiler error if anything in
+            // Self changes, which will alert us to change this function as well
+            let Self { ptr, maps, progs } = &mut self;
+            std::mem::take(maps);
+            std::mem::take(progs);
+            *ptr
+        };
+        // avoid double free of self.ptr
+        std::mem::forget(self);
```

Comment:
The `Drop` implementation of `Self` calls `libbpf_sys::bpf_object__close(self.ptr.as_ptr())`, but in this function we are taking ownership of pointer, so we must somehow avoid that `close` call being called on this ptr. This I must forget `self` in order to avoid a double free.

The rigmarole above with the `maps` and `progs` is to avoid those being leaked instead of properly dropped.

I don't like this code very much either but I couldn't find a better way. The previous implementation set `ptr` to `null` but I can't do that anymore because of `NonNull`

Reply:
Ah, makes sense. I suppose it's acceptable.